### PR TITLE
handle slice expansion

### PR DESCRIPTION
**Pull request by xh3b4sd**
I checked `fleetctl` and there is no range expansion for the slices. So `fleetctl start unit@{1..3}` is supported via the executing shell. The result is this `{"units/simple@1.service", "units/simple@2.service", "units/simple@3.service"}`. We now need to handle the arguments properly.

RFR @JosephSalisbury @zeisss

### CODE REVIEW

**Comment by zeisss** [comment]
> {"units/simple@1.service", "units/simple@2.service", "units/simple@3.service"}

So the shell also expands the subfolders?

**Comment by xh3b4sd** [comment]
The shell expands arguments. Arguments are simple strings. The example output in the description is only a example I played around with.

**Comment by xh3b4sd** [comment]
I added documentation. Please have another look.

File: docs/slice_expansion.md [hunk 1]
```diff
@@ -0,0 +1,36 @@
...
+myapp@1
+myapp@2
+myapp@3
+```
+
+Formica takes these arguments and looks up the directory `myapp`. Then the
+slice IDs `1`, `2` and `3` are used to create the unit file names accordingly.
+So lets imagine there are this two unit files within `myapp`.
+```
+some_unit_name@.service
```
Comment (zeisss) [inline docs/slice_expansion.md]: I would add the app as a prefix here, although its not 100% important to explain this feature, it might confuse people.

Comment (xh3b4sd) [inline docs/slice_expansion.md]: Done

File: cli/common.go [hunk 2]
```diff
@@ -62,7 +93,7 @@ func createRequest(dirName string) (controller.Request, error) {
 		req.Units = append(req.Units, controller.Unit{Name: name, Content: content})
 	}
 
-	sliceIDs, err := createSliceIDs(dirName)
+	sliceIDs, err := createSliceIDs(slices)
```
Comment (zeisss) [inline cli/common.go:96]: do we maybe want to return the actual groupname from `createSliceIDs` too? Our `controller.Request` is missing that, although I think we should add it there.

File: cli/common_test.go [hunk 3]
```diff
@@ -63,6 +63,29 @@ func Test_Common_createRequest(t *testing.T) {
...
+			Setup: []testFileSystemSetup{
+				{
+					FileName:    "dirname/dirname_unit@.service",
+					FileContent: []byte("some unit content"),
+					FilePerm:    os.FileMode(0644),
+				},
+			},
+			Input: []string{"dirname@1"},
+			Expected: controller.Request{
+				SliceIDs: []string{"1", "foo", "5"},
```
Comment (zeisss) [inline cli/common_test.go:80]: I don't get this test - where is `foo` and `5` coming from?

File: cli/common_test.go [hunk 4]
```diff
@@ -63,6 +63,29 @@ func Test_Common_createRequest(t *testing.T) {
 				},
 			},
 		},
+
+		// This test ensures that loading a single unit from a directory with the
+		// slice expression "@1", "@foo" and "@5" results in the expected
+		// controller request.
+		{
+			Setup: []testFileSystemSetup{
```
Comment (zeisss) [inline cli/common_test.go:71]: FYI: we could return this standard test setup into a function which we just repeat here:

```
{ 
  Setup: GivenSimpleSingleUnitFileSystemSetup(),
  [..]
```

File: cli/common_test.go [hunk 5]
```diff
@@ -30,7 +30,7 @@ func Test_Common_createRequest(t *testing.T) {
 					FilePerm:    os.FileMode(0644),
 				},
 			},
-			Input: "dirname",
+			Input: []string{"dirname"},
```
Comment (zeisss) [inline cli/common_test.go:33]: I think the `Setup` for this test shouldn't have a `@` in the unitfilename.

**Comment by zeisss** [comment]
a few minor things about the tests, but LGTM

File: cli/common.go (hunk 2 above)
Comment (xh3b4sd) [inline cli/common.go:96]: Why should we do that? For what is it needed?

File: cli/common_test.go (hunk 5 above)
Comment (xh3b4sd) [inline cli/common_test.go:33]: Fixed.

File: cli/common_test.go (hunk 4 above)
Comment (xh3b4sd) [inline cli/common_test.go:71]: Currently not a big fan of that.

File: cli/common.go (hunk 2 above)
Comment (zeisss) [inline cli/common.go:96]: mh, its probably indeed useless, but to me it feels more correct. Lets assume this is the call:

```
err := controller.Submit(controller.SubmitRequest{
  SliceIDs: []string{"1", "2", "3"},
  Units: []controller.Unit{
     controller.Unit{"appd-main@.service", "..."},
     controller.Unit{"appd-lb-register@.service", "..."},
     controller.Unit{"appd-ambassador@.service", "..."},
  }
})
```

to my eye it just feels weird to not have a `Name` field in there. Also I would probably expect the controller to double check the provided arguments (do the units start with `Name`, if we have slices, do the units contain a `@` etc.), so then its needed. 

But again: Currently its not needed and this is purely my mind wishing this for sugar reasons.

File: cli/common_test.go (hunk 3 above)
Comment (xh3b4sd) [inline cli/common_test.go:80]: The input should provide that. The fact that the test works anyway is bad. Found the bug. Fixed the test.

File: cli/common.go (hunk 2 above)
Comment (xh3b4sd) [inline cli/common.go:96]: My sugar reasoning was against having this stuff in the cli package. Anyway, lets see where we are going.

**Pull request merged**